src/mz/src/bin/mz/upgrader.rs

```diff
@@ -0,0 +1,289 @@
+
+        if major > local_version.major {
+            return true;
+        }
+
+        if minor > local_version.minor {
+            return true;
+        }
+
+        if major == local_version.major
+            && minor == local_version.minor
+            && patch > local_version.patch + 5
+        {
+            return true;
+        }
+
+        false
+    }
+
+    /// Fetches and returns the latest tag version from the Materialize repository
+    /// using the GitHub public API.
+    async fn get_last_tag_version(&self) -> Result<Option<(u64, u64, u64)>, Error> {
+        #[derive(Deserialize, Debug)]
+        struct GithubTag {
+            name: String,
+        }
+
+        // Headers recommended by GitHub: https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-repository-tags
+        let mut headers = HeaderMap::new();
+        headers.insert(
+            ACCEPT,
+            HeaderValue::from_static("application/vnd.github+json"),
+        );
+        // The GitHub API endpoint requires setting the user agent; otherwise, it returns a 403.
+        let user_agent = format!("mz/{}", VERSION.clone());
+        headers.insert(
+            USER_AGENT,
+            HeaderValue::from_str(&user_agent).map_err(Error::HeaderParseError)?,
+        );
+
+        let client = reqwest::Client::new();
+
+        // As of the creation date of this code, the last `mz-vx.y.z` tag is located on page 4 (last-one).
+        // I kept the first page in the event to avoid any issues.
+        let mut page = 1;
+
+        let last_tag = loop {
+            let url = format!(
+                "https://api.github.com/repos/materializeInc/materialize/tags?per_page=100&page={}",
+                page
+            );
+            let response = client
+                .get(url)
+                .headers(headers.clone())
+                .send()
+                .await
+                .map_err(Error::GitHubFetchError)?;
+
+            let tags: Vec<GithubTag> = response
+                .json()
+                .await
+                .map_err(Error::ReqwestJsonParseError)?;
+
+            if let Some(tag) = tags
+                .into_iter()
+                .find(|tag| tag.name.starts_with(MZ_V_PREFIX))
+            {
+                break tag;
+            }
+            page += 1;
+        };
+
+        let last_version = self.strip_version(last_tag.name)?;
+
+        Ok(last_version)
+    }
+
+    /// Formats and returns the update message depending on the user's operating system.
+    fn format_version_update_message(&self) -> String {
+        let os = std::env::consts::OS;
+
+        match os {
+            "linux" => {
+                "New version available. Update now: `apt install --only-upgrade materialize-cli`"
+            }
+            "macos" => {
+                "New version available. Update now: `brew upgrade materializeInc/materialize/mz`"
+            }
+            _ => "New version available. Update now.",
```

Comment:
Hrm, this isn't going to work for folks who installed `mz` by curling the binary. I think we probably want to do something a little fancier: have the apt and Homebrew packages _indicate_ to `mz` that apt/Homebrew were used to install `mz`. 

Specifically, I'd have apt write `apt` into `/usr/share/mz/install-method` and Homebrew write `brew` into `/opt/homebrew/opt/mz/share/install-method`.

Then, `mz` can determine how it was installed by calling [`std::env::current_exe`](https://doc.rust-lang.org/std/env/fn.current_exe.html), traversing to "../../share/install-method", and reading that file's contents. If the file doesn't exist, we can assume `curl`, and provide `curl` instructions. If the file does exist, we can look inside to determine whether the install method was `apt` or `brew`, and provide `apt` or `brew` instructions, accordingly.

Reply:
One of the previous approaches, maybe the simplest one, was to just display: `New version available: v0.1.2. Please update now.` It was less informative about the update process but much less prone to incorrect outputs.